### ISSUE TITLE
db.bolt

### ISSUE THREAD

**Issue by 0323pin** [elves/elvish#1621]
Hi,

I've been using `elvish` as my login shell for about 6 months. Every now and then, I clear the history by deleting `~/.local/state` directory. Up-to-now, this was simply regenerated but, since a week or so, it no longer does.

I wonder if this has anything to do with the change of `go` default version on NetBSD from 1.18 to 1.19 but, I wouldn't think so. This change was pushed nearly 2 months ago and, my issue is much more recent.

Anyway, is there a command that I could use to generate a new `~/.local/state/elvish/db.bolt`?
Thanks.

**Comment by krader1961** [comment]
Elvish launches a daemon to manage the I/O to the database. The daemon keeps the database file open as long as it is running. When you remove the database file the daemon keeps doing I/O to that, now nameless, file. You need to kill the Elvish daemon process to force a new daemon to be launched which will create a new db.bolt file. See also the [store](https://elv.sh/ref/store.html) module.

**Comment by 0323pin** [comment]
Thank you for your reply.

> You need to kill the Elvish daemon process to force a new daemon to be launched which will create a new db.bolt file.

But, I've even rebooted my system and nothing. The history is kept, in memory I guess, while running a terminal instance but, forgotten as soon as I close the terminal and launch a new one.

I'll try to change shell back to `ksh` and then back to `elvish` to see if it helps. Weird that this didn't happen for 6 months and now I'm having this issue.

**Comment by krader1961** [comment]
Possibly you've changed where the db file lives or accidentally created the database at the legacy path by running an old version of Elvish. See https://elv.sh/ref/command.html#database-file.

> The history is kept, in memory I guess, while running a terminal instance but, forgotten as soon as I close the terminal and launch a new one.

That isn't correct. It is persisted to disk. Also, the daemon continues to run until there are no interactive elvish processes.

**Comment by 0323pin** [comment]
> accidentally created the database at the legacy path by running an old version of Elvish

I haven't used anything else other than,
```
~ > elvish -version
0.18.0
```

> Possibly you've changed where the db file lives

Could be. Is there a way to find out?
Because `~/.elvish/db` and `$XDG_STATE_HOME/elvish/db.bolt` do not exist.

**Comment by xiaq** [comment]
The database should be created with only rw permission to the owner. This is what I have:

```elvish-transcript
~> ls -l ~/.local/state/elvish/db.bolt
-rw-------  1 xiaq  staff  4194304 11 14 16:18 /Users/xiaq/.local/state/elvish/db.bolt
```

However, Elvish doesn't check the permission of existing files, so if the permission is not correct it should still be fine.

Creating an empty file is not going to work because an empty file won't be a valid BoltDB database.

Questions to you:

- Is there any other symptom? If there's anything wrong with the storage there should usually be an error message during startup.

- How did you install Elvish? Does the problem exist if you use an official binary from https://elv.sh/get/?

**Comment by xiaq** [comment]
If you are not seeing any error messages, I have a suspicion that you are using a build of the `nodaemon` variant. In that case using an official binary should fix the problem.

**Comment by 0323pin** [comment]
@xiaq 

> Is there any other symptom? If there's anything wrong with the storage there should usually be an error message during startup.

No, no other symptoms, everything is working as before. There's most probably nothing wrong with storage. Let me see if I can explain it in another way,

![2022-11-14-175747_1366x768_scrot](https://user-images.githubusercontent.com/90570748/201721213-4ce80ac7-c32b-4555-adaa-9cc01dd3e7be.png)

In the screenshot, two instances of `alacritty` running. In one I've typed three commands followed by Ctrl+r, as you can see history is displayed. In the second one, history is empty. If I close both and launch a new terminal, history is empty and there's no `.local/state/elvish/db.bolt` file.

> How did you install Elvish? Does the problem exist if you use an official binary from https://elv.sh/get/?

I'm a NetBSD dev and build it from source. Actually, I'm usually the one updating the package.
Here's the Makefile from pkgsrc,

```
# $NetBSD: Makefile,v 1.21 2022/11/02 19:39:51 bsiegert Exp $

DISTNAME=	elvish-0.18.0
PKGREVISION=	8
CATEGORIES=	shells
MASTER_SITES=	${MASTER_SITE_GITHUB:=elves/}
GITHUB_TAG=	v${PKGVERSION_NOREV}

MAINTAINER=	mj@mjturner.net
HOMEPAGE=	https://elv.sh/
COMMENT=	Expressive programming language and a versatile interactive shell
LICENSE=	2-clause-bsd

INSTALLATION_DIRS=	bin
PKG_SHELL=		bin/elvish
GO_BUILD_PATTERN=	./...

# needed for the tests
USE_TOOLS+=		perl gmake

do-install:
	${INSTALL_PROGRAM} ${WRKDIR}/.gopath/bin/elvish \
		${DESTDIR}${PREFIX}/bin

.include "go-modules.mk"

.include "../../lang/go/go-module.mk"
.include "../../mk/bsd.pkg.mk"
```
> I have a suspicion that you are using a build of the nodaemon variant. In that case using an official binary should fix the problem.

As said, I've been running `elvish` for about 6 months as my default shell and this issue only started a week or so ago. Before that everything was fine ad the database was regenerated whenever I deleted it.
Also, as you can see from the Makefile, we don't use the nodeamon flag.

Moreover, `ps` shows `2727 ttyE0 Il+  0:00.09 -elvish` so, it seems to be running.

Honestly, this is very surprising and I wouldn't open an issue otherwise. I was hoping that it would be possible to check where the database is located, in case @krader1961 is right about it being moved somewhere else or, that there was a simple way to force the generation of a new database.

I've looked through my home dir several times but, I simply can't find anything wrong.

```
/home > find pin -name elvish
pin/Git/leftwm-netbsd/.config/elvish
pin/.config/elvish
```

**Comment by 0323pin** [comment]
@xiaq and @krader1961 As suspected something was wrong but, I couldn't figure out exactly what. So, I logged in as root, removed `elvish` and rebuild it from source with go-1.19.3. Then, as root I've removed nearly everything from `/home/pin`, just left my git-repo and `.gitconfig`.

After, I've restored the config files from my repository and `chown`ed `/home/pin` to adjust permissions. Logged out and back in as user  _et voilà_ ...

```
~ > ls .local/state/elvish/
.rw------- pin wheel 32 KB Mon Nov 14 19:18:51 2022  db.bolt
```

History is working as previously.
